imapclient/tls.py

```diff
@@ -5,200 +5,30 @@
-        context.options |= getattr(_ossl, "OP_SINGLE_DH_USE", 0)
-        context.options |= getattr(_ossl, "OP_SINGLE_ECDH_USE", 0)
-
-        # disallow ciphers with known vulnerabilities
-        # TODO: backports.ssl.SSLContext is missing set_ciphers
-        context._ctx.set_cipher_list(_RESTRICTED_SERVER_CIPHERS)
-
-        if cafile or capath or cadata:
-            context.load_verify_locations(cafile, capath, cadata)
-        elif context.verify_mode != ssl.CERT_NONE:
-            # no explicit cafile, capath or cadata but the verify mode is
-            # CERT_OPTIONAL or CERT_REQUIRED. Let's try to load default system
-            # root CA certificates for the given purpose. This may fail silently.
-            if sys.platform == "win32":
-                certs = bytearray()
-                for storename in ("CA", "ROOT"):
-                    for cert, encoding, trust in enum_certificates(storename):
-                        # CA certs are never PKCS#7 encoded
-                        if encoding == "x509_asn":
-                            if trust is True or Purpose.SERVER_AUTH in trust:
-                                certs.extend(cert)
-
-                if certs:
-                    context.load_verify_locations(cadata=certs)
-            else:
-                context.set_default_verify_paths()
-
-        return context
-
-
-    def _wrap_socket(sock, ssl_context, hostname):
-        """Wrap a socket and return an SSLSocket.
-
-        If *ssl_context* is `None`, a default context as returned by
-        `create_default_context` will be used.
-
-        If certificate validation fails, the socket will be shut down and
-        an Error raised.
-        """
-        if not ssl_context:
-            ssl_context = create_default_context()
-
-        def killsock():
-            sock.shutdown(socket.SHUT_RDWR)
-            sock.close()
-
-        try:
-            newsock = ssl_context.wrap_socket(sock, server_hostname=hostname)
-        except ssl.CertificateError as err:
-            killsock()
-            raise imaplib.IMAP4.error("certificate error for %s: %s" % (hostname, str(err)))
-        except ssl.SSLError as err:
-            killsock()
-            raise imaplib.IMAP4.error("SSL error for %s: %s" % (hostname, err.args[-1]))
-
-        return _SSLSocketWithShutdown(newsock)
-
-    wrap_socket = lambda sock, context, host: _wrap_socket(sock, context, host)
-
-
-    # TODO: get shutdown added in backports.ssl.SSLSocket
-    class _SSLSocketWithShutdown(object):
-
-        def __init__(self, sslsock):
-            self.sslsock = sslsock
+def create_default_context(cafile=None, capath=None, cadata=None):
+    """TODO: doc
+    """
+    return ssl.create_default_context(purpose=ssl.Purpose.CLIENT_AUTH,
+                                      cafile=cafile,
+                                      capath=capath,
+                                      cadata=cadata)
```

Comment:
We should make sure that this context actually works with the big IMAP providers. TLS provided by IMAP servers is usually somewhat behind what is offered by web servers in terms of TLS versions, ciphers etc.

Reply:
We use the default context to connect to the big players without any issue.

Reply:
Ok, that's great. 

The only reason to keep this is as a convenience so you don't have to set the CLIENT_AUTH purpose, but that's probably not reason enough to keep it. I'm happy for it to go.

There may be examples in the docs which need to be updated.